### ISSUE TITLE
Tensorpack conv2d seems output different result with self implemented conv layer 

### ISSUE THREAD

**Issue by FCInter** [tensorpack/tensorpack#1485]
I'm trying to convert a `resnet101` model checkpoint from tensorpack to `pytorch`. The checkpoint file has been converted but the inference results do not match. The key problem is that, from the very beginning, i.e. the first `conv2d` layer, the results become different.

After finding this problem, I implemented a conv layer using `numpy ` only. Its result matches very well with `pytorch`. But it's also significantly different with `tensorpack`. However, what's more embarrassing is that, the network output, i.e. the predicted label of an input image, is wrong in `pytorch`, but correct in `tensorpack`.

In summary:

1. I got seemly wrong output of conv2d in `tensorpack`, but in the meanwhile got correct network output.

2. I got seemly correct output of conv2d in `pytorch` and `numpy`, but in the meanwhile got wrong network output. 

The following is my `tensorpack` code (to make it short, I only post the key parts):

tensorpack model:

    class Model(ImageNetModel):
        def __init__(self, depth, data_format='NCHW', mode='resnet'):
            super(Model, self).__init__(data_format)
    
            if mode == 'se':
                assert depth >= 50
    
            self.mode = mode
            basicblock = preresnet_basicblock if mode == 'preact' else resnet_basicblock
            bottleneck = {
                'resnet': resnet_bottleneck,
                'preact': preresnet_bottleneck,
                'se': se_resnet_bottleneck}[mode]
            self.num_blocks, self.block_func = {
                18: ([2, 2, 2, 2], basicblock),
                34: ([3, 4, 6, 3], basicblock),
                50: ([3, 4, 6, 3], bottleneck),
                101: ([3, 4, 23, 3], bottleneck),
                152: ([3, 8, 36, 3], bottleneck)
            }[depth]
    
        def get_logits(self, image):
            with argscope([Conv2D, MaxPooling, GlobalAvgPooling, BatchNorm], data_format=self.data_format):
                return resnet_backbone(
                    image, self.num_blocks,
                    preresnet_group if self.mode == 'preact' else resnet_group, self.block_func)

The `resnet_backbone` of the model is :

    def resnet_backbone(image, num_blocks, group_func, block_func):
        with argscope(Conv2D, nl=tf.identity, use_bias=False,
                      W_init=tf.variance_scaling_initializer(scale=2.0, mode='fan_out')):
            l1 = (LinearWrap(image).Conv2D('conv0', 64, 7, stride=2, nl=tf.identity)())
            logits = (LinearWrap(l1)
                .BatchNorm('conv0/bn')
                .apply(tf.nn.relu)
                .MaxPooling('pool0', shape=3, stride=2, padding='SAME')
                .apply(group_func, 'group0', block_func, 64, num_blocks[0], 1)
                .apply(group_func, 'group1', block_func, 128, num_blocks[1], 2)
                .apply(group_func, 'group2', block_func, 256, num_blocks[2], 2)
                .apply(group_func, 'group3', block_func, 512, num_blocks[3], 2)
                .GlobalAvgPooling('gap')
                .FullyConnected('linear', NUM_CLASSES, nl=tf.identity)())
        return logits, l1

In particular, I return the `l1` tensor, which is the output of the first conv layer of the resnet. Then I compare it with the output of the `pytorch ` conv layer.

The pytorch model is directly imported from the `torchvision.models`, and also I return the output of the first conv layer.

The input image is of size (1,3,224,224), the output shape is (1, 64, 112, 112). Both results match. 

Then I made 3 tests:

1. The values of two results are quite different. The summation of absolute different is in scale of 1e6.

2. I implement a conv layer using `numpy`. Its result matches pretty well with `pytorch`, total difference < 0.1. And it's also 1e6 different from `tensorpack`.

3. In `tensorpack` code, I also tried to return the `image` (i.e. the input) directly to the output. And compare it with the input of `pytorch`. The two inputs match in both shape and value. So I'm quite sure the input and kernel values of the two `conv2d` are exactly the same.

What's wrong with my implementation?

Thank you all for helping me!

**Comment by ppwwyyxx** [comment]
Tensorpack's convolution is correct as long as TensorFlow's convolution is correct. Tensorpack does not implement convolution.

PyTorch and TensorFlow's definition of convolution padding is different. Getting unexpected results out of the box is expected.

Closing as the issue is unrelated to tensorpack.
